src/sky/SkyColoring.cpp

```diff
@@ -0,0 +1,369 @@
+#include "SkyColoring.hpp"
+#include "SkyConfig.h"
+#include <Image/BsColor.h>
+#include <Math/BsMath.h>
+#include <Renderer/BsCamera.h>
+#include <Scene/BsSceneManager.h>
+
+namespace REGoth
+{
+  const float TIME_KEY_0 = 0.00f;
+  const float TIME_KEY_1 = 0.25f;
+  const float TIME_KEY_2 = 0.30f;
+  const float TIME_KEY_3 = 0.35f;
+  const float TIME_KEY_4 = 0.50f;
+  const float TIME_KEY_5 = 0.65f;
+  const float TIME_KEY_6 = 0.70f;
+  const float TIME_KEY_7 = 0.75f;
+
+  SkyColoring::SkyColoring()
+  {
+    
+  }
+
+  SkyColoring::~SkyColoring()
+  {
+  }
+
+  void SkyColoring::getSkyColors(bs::Color& color0, bs::Color& color1)
+  {
+    color0 = bs::Color(0.0f, 0.0f, 0.0f, 0.0f);
+    color1 = mMasterState.baseColor;
+  }
+
+  void SkyColoring::interpolate(float dayRatio)
+  {
+    mMasterState.time = std::fmod(dayRatio + 0.5, 1.0);
+
+    // init with values for case: time >= TIME_KEY_7 (= 0.75f)
+    bs::UINT32 si0 = (bs::INT32)mSkyStates.size() - 1;
+    bs::UINT32 si1 = 0;
+
+    // Find the two states we're interpolating
+    for (bs::UINT32 i = 0; i < (bs::UINT32)mSkyStates.size(); i++)
+    {
+      // Interpolate between the current and the next
+      if (mMasterState.time < mSkyStates[i].time)
+      {
+        si0 = (i - 1) % mSkyStates.size();
```

Comment:
Are you sure, doing `i - 1` if `i = 0` is a good idea? Also is the intent to interpolate between the previous and the next state (that would be strange) or the current and the next state as stated in the comment?

Reply:
That looks weird indeed, but it is correct.

For example: Let's say the master states time is `0.30` and we got state A at `0.00` and state B at `0.25` and state C at `0.50`. Since they are ordered, we would need to find the first state with a time larger than the current time. In this case, that would be state C. Of course we overshot by one, since we're interested in state B, that's where the subtraction comes from.
As long as they're set up correctly, the case that `i` will be 0 will not happen.

I agree that all this should be made more clear. Finding the first state with a time *less* than the current time would also work and be much more readable.

Reply:
Thanks for the additional clarification!

Especially I didn't think of this when making my comment: 

> Subtracting 1 will not cause a negative numbers here **since the first state has a time of 0.0**